src/main/java/com/lginterview/providers/CourseProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+        this.userRepository = userRepository;
+
+        this.courseActions.put(RequestType.INIT, (this::createCourse));
+        this.courseActions.put(RequestType.SAVE, (this::updateCourse));
+        this.courseActions.put(RequestType.FINISH, (this::closeCourse));
+    }
+
+    @Override
+    @Transactional
+    public Long modifyCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        if (courseRequest == null)
+            throw new IllegalArgumentException("Course Request cannot be null");
+
+        if (courseRequest.getType() == null)
+            throw new IllegalArgumentException("Course Request Type cannot be null");
+
+        if (courseActions.containsKey(courseRequest.getType()))
+            return courseActions.get(courseRequest.getType()).apply(courseRequest);
+        else
+            throw new IllegalArgumentException("Invalid Course Request type");
+    }
+
+    private Long createCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course newCourse = courseRepository.findOne(courseRequest.getCourseId());
+        if (newCourse != null)
+            return newCourse.getId();
+
+        User courseUser = userRepository.findOne(courseRequest.getUserId());
+        if (courseUser == null)
+            throw new CourseDataRetrievalException("User not found");
+
+        newCourse = new Course();
+        newCourse.setStartDate(courseRequest.getTimestamp());
+        newCourse.setCourseUser(courseUser);
+        return courseRepository.save(newCourse).getId();
+    }
+
+    private Long updateCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+
+        Course courseToUpdate = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToUpdate == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        if (courseRequest.getScore() != null) {
+            int scoreValue = courseRequest.getScore();
+            courseToUpdate.getScores().add(scoreValue);
+            courseRepository.save(courseToUpdate);
+        }
+        return courseToUpdate.getId();
+    }
+
+    private Long closeCourse(CourseRequest courseRequest) throws CourseDataRetrievalException {
+        Course courseToClose = courseRepository.findOne(courseRequest.getCourseId());
+        if (courseToClose == null)
+            throw new CourseDataRetrievalException("Course Not Found");
+
+        courseToClose.setEndDate(courseRequest.getTimestamp());
+        courseRepository.save(courseToClose);
+        return courseToClose.getId();
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public CourseStatus getCourseStatus(long userId, long courseId) throws CourseDataRetrievalException {
+
+        Course courseData = getCourse(userId, courseId);
+
+        int bestScore = 0;
+        int averageScore = 0;
+
+        if ((courseData.getScores() != null)) {
+            bestScore = courseData.getScores()
```

Comment:
[MS][Major] In my opinion Course class should have a method Course::bestScore(), Course::averageScore() and Course::courseStats().

See https://martinfowler.com/bliki/TellDontAsk.html.

Reply:
I Agree